algorithms/rainbow/model.py

```diff
@@ -0,0 +1,112 @@
+"""
+Model definition definition for DQN and Noisy layers
+"""
+import math
+import torch
+from torch import nn
+from torch.nn import functional as F
+
+
+class RainbowDQN(nn.Module):
+    """
+    From the paper "Rainbow: Combining Improvements in Deep Reinforcement Learning"
+    Source: https://arxiv.org/pdf/1710.02298.pdf
+
+    Main model used as Q function estimator for policy generation
+    """
+    def __init__(self, args, action_space):
+        super().__init__()
+        self.atoms = args.atoms
+        self.action_space = action_space.n
+        self.linear_in = self.get_linear_size(args)
+
+        self.conv1 = nn.Conv2d(args.history_length, 32, 8, stride=4, padding=1)
+        self.conv2 = nn.Conv2d(32, 64, 4, stride=2)
+        self.conv3 = nn.Conv2d(64, 64, 3)
+        self.fc_h_v = NoisyLinear(self.linear_in, args.hidden_size, std_init=args.noisy_std)
+        self.fc_h_a = NoisyLinear(self.linear_in, args.hidden_size, std_init=args.noisy_std)
+        self.fc_z_v = NoisyLinear(args.hidden_size, self.atoms, std_init=args.noisy_std)
+        self.fc_z_a = NoisyLinear(args.hidden_size, self.action_space * self.atoms, std_init=args.noisy_std)
+
+    def forward(self, x, log=False):
+        x = F.relu(self.conv1(x))
+        x = F.relu(self.conv2(x))
+        x = F.relu(self.conv3(x))
+        x = x.view(-1, self.linear_in)
+        v = self.fc_z_v(F.relu(self.fc_h_v(x)))  # Value stream
+        a = self.fc_z_a(F.relu(self.fc_h_a(x)))  # Advantage stream
+        v, a = v.view(-1, 1, self.atoms), a.view(-1, self.action_space, self.atoms)
```

Comment:
this magic view puts everything into the right shape atom wise and we STILL do q,v a decomposition on the atoms?!

Reply:
x = features after conv net. Then each stream tries to figure out its own atoms which will be merged into Q. What should I change here?

Reply:
add a comment that we have the value and advantage DISTRIBUTIONS with the atoms reshaped here

Reply:
view is just the equivalent of numpy.reshape for torch.Tensor variables. Changed naming for clarity, but I don't think it needs much more explanation

Reply:
I know what view is?
I wanted a comment to clearly state that this isn't just Q+V decomposition, it's Q+V distribution decomposition...